feat(template): introduce rxChunk

hoebbelsB opened the pull request:
# ChunkDirective

This PR introduces the `*rxChunk` structural directive.

the  `*rxChunk` directive serves as a convenient way for dividing template work into chunks.
Applied to an element, it will schedule a task with the given RxRenderStrategy in order to postpone the template creation of this element.

![rx-chunk](https://user-images.githubusercontent.com/4904455/180814988-14887411-9106-4ee7-9e0d-22a46e127b66.gif)

closes https://github.com/rx-angular/rx-angular/issues/1802

### PR Checklist

- [x] implementation
- [x] demo
- [x] unit tests
- [x] docs (jsdocs)

### Features of `*rxChunk`

* lightweight alternative to `*rxLet="[]"`
* renderCallback
* no value binding
* no context variables
* zone agnostic
* suspense template

![long task](https://user-images.githubusercontent.com/4904455/180815459-1d49aa72-97fa-42b4-bccb-437a42f8d9b0.png)

![distribute work](https://user-images.githubusercontent.com/4904455/180815471-cd31a234-1fa9-487e-bba2-60735bd9591e.png)

Karnaukhov-kh commented on libs/template/chunk/src/lib/chunk.directive.ts:
```diff
@@ -0,0 +1,174 @@
+   * @param {NextObserver<void>} renderCallback
+   */
+  @Input('rxChunkRenderCallback')
+  set renderCallback(renderCallback: NextObserver<void>) {
```
Is it possible to use not only subjects but also just a function? I know it's more like a general question for render callbacks but would it worth the try?

hoebbelsB replied:
I like the idea, I would even more like to just have an `@Output`. But it looks like it's not supported, at least for the `*` syntax. https://github.com/angular/angular/issues/12121

Karnaukhov-kh commented on libs/template/chunk/src/lib/chunk.directive.ts:
```diff
@@ -0,0 +1,174 @@
+   *
+   * @param { RxStrategyNames<string> } strategy
+   */
+  @Input('rxChunk') strategy = this.strategyProvider.primaryStrategy;
```
Is it planned to work with not concurrent strategies?

hoebbelsB replied:
sure, but why would u? it basically makes the `rxChunk` do nothing useful anymore :D

Karnaukhov-kh replied:
Yes, that's the thing. Can we type this input to RxConcurrentStrategyNames?

BioPhoton replied:
We have to support all Strategien. Other strategies then concurrent will have an effect definitely. U could e.g. use 'noop' strategy when out of viewport

ChristopherPHolder replied:
There is a need to support all strategies. If you want to change the rendering strategy conditionally, then this is necessary. 

And it makes sense in many cases to switch the strategies based on conditions like first render and other things which may affect CLS

Karnaukhov-kh commented on libs/template/chunk/src/lib/tests/chunk.directive.spec.ts:
```diff
@@ -0,0 +1,82 @@
+    ['immediate'],
+    ['userBlocking'],
+    ['normal'],
+    ['low'],
```
If we plan to work with not concurrent strats, should we test it?

BioPhoton replied:
Yes we should test normal too

hoebbelsB replied:
done

Karnaukhov-kh reviewed (approved):
Hi @hoebbelsB really like the idea. Looks super interesting!

edbzn commented:
@hoebbelsB shouldn't we wait for the first stable release before introducing new elements?

BioPhoton commented:
As not all strategies chunk let's call it rxTemplate?

ChristopherPHolder commented:
Hi, the current workaround for this is to do *rxLet="[]" 

Which from anyone who is not aware of the pattern makes 0 sense. 

Is there anything blocking the issue? 

Is it only the name that's blocking? 

`*rxChunk` - `*rxTemplate` - `*rxRender`

AdrianRomanski commented:
@Karnaukhov-kh @hoebbelsB  Everything is looking solid :) 

***rxChunk** is best name in my opinion - referring to Christopher last comment. 

That feature have a great marketing potential, to be announced or either be shown in a blog post. 
Lets go and merge it :D

ChristopherPHolder commented:
@Karnaukhov-kh @hoebbelsB @AdrianRomanski 

Just noticed that this MR has been open since before standalone 😂

I would be happy to clean that up but i guess its kind of clear that there is not plans to merge this?

hoebbelsB committed:
feat(template): introduce rxChunk directive

RxChunk is a shortcut for `*rxLet="[]"`. It also has the ability render a suspense view as long the actual view isn't rendered yet.

hoebbelsB committed:
feat(demos): implement rxChunk demo

eneajaho commented on libs/template/chunk/src/lib/chunk.directive.ts:
```diff
@@ -0,0 +1,178 @@
+   *
+   * @param { boolean } patchZone
+   */
+  @Input('rxChunkPatchZone') patchZone =
```
Do we still want this one to be at input level, instead of having it at the provider level?

hoebbelsB replied:
well, it's anyway on provider level. I guess we can get rid of it

hoebbelsB replied:
we could get rid of it, yeah
